File: CashFlow/Models/DataModelDateUtils.m
// -*-  Mode:ObjC; c-basic-offset:4; tab-width:8; indent-tabs-mode:nil -*-
/*
 * CashFlow for iOS
 * Copyright (C) 2008-2011, Takuya Murakami, All rights reserved.
 * For conditions of distribution and use, see LICENSE file.
 */

#import "CashFlow-Swift.h"

#import <Foundation/Foundation.h>
#import "DataModel.h"
#import "DataModelDateUtils.h"

@implementation DataModel(DateUtils)

//
// DateFormatter
//

+ (NSDateFormatter *)dateFormatter
{
    static NSDateFormatter *dfDateOnly = nil;
    static NSDateFormatter *dfDateTime = nil;
    
    if ([Config instance].dateTimeMode == DateTimeModeDateOnly) {
        if (dfDateOnly == nil) {
            dfDateOnly = [self dateFormatter:NSDateFormatterNoStyle withDayOfWeek:YES];
        }
        return dfDateOnly;
    } else {
        if (dfDateTime == nil) {
            dfDateTime = [self dateFormatter:NSDateFormatterShortStyle withDayOfWeek:YES];
        }
        return dfDateTime;
    }
}

+ (NSDateFormatter *)dateFormatter:(BOOL)withDayOfWeek
{
    if ([Config instance].dateTimeMode == DateTimeModeDateOnly) {
        return [self dateFormatter:NSDateFormatterNoStyle withDayOfWeek:withDayOfWeek];
    } else {
        return [self dateFormatter:NSDateFormatterShortStyle withDayOfWeek:withDayOfWeek];
    }
}

+ (NSDateFormatter *)dateFormatter:(NSDateFormatterStyle)timeStyle withDayOfWeek:(BOOL)withDayOfWeek
{
    NSDateFormatter *df = [NSDateFormatter new];
    df.dateStyle = NSDateFormatterMediumStyle;
    df.timeStyle = timeStyle;
    
    NSMutableString *s = [NSMutableString stringWithCapacity:30];
    [s setString:df.dateFormat];
    
    if (withDayOfWeek) {
        [s replaceOccurrencesOfString:@"MMM d, y" withString:@"EEE, MMM d, y" options:NSLiteralSearch range:NSMakeRange(0, s.length)];
        [s replaceOccurrencesOfString:@"yyyy/MM/dd" withString:@"yyyy/MM/dd(EEEEE)" options:NSLiteralSearch range:NSMakeRange(0, s.length)];
    }
    
    df.dateFormat = s;
    return df;
}

@end

File: CashFlow/Models/DataModelDateUtils.swift
/*
 * CashFlow for iOS
 * Copyright (C) 2008-2015, Takuya Murakami, All rights reserved.
 * For conditions of distribution and use, see LICENSE file.
 */

import Foundation

extension DataModel {
    private static var dfDateOnly: NSDateFormatter?
    private static var dfDateTime: NSDateFormatter?
    
    static func dateFormatter() -> NSDateFormatter {
        if (Config.instance().dateTimeMode == .DateOnly) {
            if (dfDateOnly == nil) {
                dfDateOnly = DataModel.dateFormatter(.NoStyle, dayOfWeek:true)
            }
            return dfDateOnly!
        } else {
            if (dfDateTime == nil) {
                dfDateTime = DataModel.dateFormatter(.ShortStyle, dayOfWeek:true)
            }
            return dfDateTime!
        }
    }

    static func dateFormatter(dayOfWeek: Bool) -> NSDateFormatter {
        if (Config.instance().dateTimeMode == .DateOnly) {
            return DataModel.dateFormatter(.NoStyle, dayOfWeek:dayOfWeek)
        } else {
            return DataModel.dateFormatter(.ShortStyle, dayOfWeek: dayOfWeek)
        }
    }

    private static func dateFormatter(timeStyle: NSDateFormatterStyle, dayOfWeek: Bool) -> NSDateFormatter {
        let df = NSDateFormatter()
        df.dateStyle = .MediumStyle
        df.timeStyle = timeStyle;
    
        var s = df.dateFormat
    
        if (dayOfWeek) {
            s = s.stringByReplacingOccurrencesOfString("MMM d, y", withString: "EEE, MMM d, y")
            s = s.stringByReplacingOccurrencesOfString("yyyy/MM/dd", withString: "yyyy/MM/dd(EEEEE)")
        }
    
        df.dateFormat = s;
        return df;
    }
}
